presto-hive/src/main/java/io/prestosql/plugin/hive/rubix/RubixInitializer.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.rubix;
+
+import com.codahale.metrics.MetricRegistry;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.qubole.rubix.bookkeeper.BookKeeper;
+import com.qubole.rubix.bookkeeper.BookKeeperServer;
+import com.qubole.rubix.bookkeeper.LocalDataTransferServer;
+import com.qubole.rubix.core.CachingFileSystem;
+import io.airlift.log.Logger;
+import io.prestosql.plugin.hive.ConfigurationInitializer;
+import io.prestosql.spi.Node;
+import io.prestosql.spi.NodeManager;
+import org.apache.hadoop.conf.Configuration;
+
+import javax.inject.Inject;
+
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/*
+ * Responsibilities of this initializer:
+ * 1. Lazily setup RubixConfigurationInitializer with information about master when it is available
+ * 2. Start Rubix Servers.
+ * 3. Inject BookKeeper object into CachingFileSystem class
+ */
+public class RubixInitializer
+{
+    private static final Logger log = Logger.get(RubixInitializer.class);
+
+    private final RubixConfigurationInitializer rubixConfigurationInitializer;
+    private final Set<ConfigurationInitializer> configurationInitializers;
+
+    @Inject
+    public RubixInitializer(RubixConfigurationInitializer rubixConfigurationInitializer, Set<ConfigurationInitializer> configurationInitializers)
+    {
+        this.rubixConfigurationInitializer = rubixConfigurationInitializer;
+        this.configurationInitializers = configurationInitializers;
+    }
+
+    public void initializeRubix(NodeManager nodeManager)
+    {
+        ExecutorService initializerService = Executors.newSingleThreadExecutor();
+        ListenableFuture<Boolean> nodeJoinFuture = MoreExecutors.listeningDecorator(initializerService).submit(() ->
+        {
+            while (!(nodeManager.getAllNodes().contains(nodeManager.getCurrentNode()) &&
```

Comment:
Can we do this via `InternalNodeManager#addNodeChangeListener` rather than the looping here ?

Reply:
Yes, It would be cleaner.
It will need changes to `NodeManager` interface and implementations to work with `InternalNodeManager's` listener methods. Handling removeNodeChangeListener in NodeManager could be tricky since callers are not aware of `AllNodes` class.